Upgrade WordPress monorepo

jsnajdr opened the pull request:
A fresh attempt to land changes from #46223.

Upgrades all `@wordpress/*` packages except `@wordpress/scripts`. That package triggers a Babel update, too, and causes too many changes in `yarn.lock`. We can do it in a separate PR.

The upgrades packages are deduplicated.

Contains all the SASS color name changes from #46223 caused by `@wordpress/base-styles` upgrade, squashed into one commit.

Contains my unit test fixes and also @sirreal's `react-i18n` typing upgrades.

@sirreal one thing I'm not 100% sure about: the new `@wordpress/*` packages all have published types. Does that mean we can remove all `@types/wordpress__*` dependencies now? They should be redundant.

Fixes #48585

renovate-bot committed:
fix(deps): update wordpress monorepo

scinos committed:
Use new colors

jsnajdr committed:
Tests: update names of internal data control types

jsnajdr committed:
Tests: fix selector resolution tests by waiting for one more state update

jsnajdr committed:
Fix SftpCard tests by mocking window.matchMedia

sirreal committed:
Type with hooks

sirreal commented:
> one thing I'm not 100% sure about: the new `@wordpress/*` packages all have published types. Does that mean we can remove all `@types/wordpress__*` dependencies now? They should be redundant.

Yep, they can be removed for any packages that include types. DefinitelyTyped has a process for removing types for packages that no longer require them which I usually do periodically, but maybe @saramarcondes would like to help as well.

jsnajdr committed:
Fix Editing Toolkit build

jsnajdr committed:
Remove types/wordpress packages

jsnajdr commented:
> Yep, they can be removed for any packages that include types.

OK, I removed the type packages, except `@types/wordpress__compose` which is not natively typed yet. Let's see what CI says.

jsnajdr commented:
The `editing-toolkit` and `wpcom-block-editor` apps needed some fixups:

There is no longer a `$big-font-size` SASS variable defined in `@wordpress/base-styles` (removed during cleanup in https://github.com/WordPress/gutenberg/pull/26827), so I replaced that with raw `18px` value. The style is used in the nav sidebar:

<img width="275" alt="Screenshot 2021-01-08 at 12 39 17" src="https://user-images.githubusercontent.com/664258/104011552-96237e80-51ae-11eb-9199-3beaa5202a8d.png">

The `__experimentalMainDashboardButton` export has also been moved (by @vindl) from `@wordpress/interface` to `@wordpress/edit-post`, so I updated the imports and dependencies accordingly.

@noahtallen @vindl What is a good way to test the upgrade here and can you possibly help? I don't really know what I'm doing when touching these parts of the codebase.

jsnajdr commented:
> OK, I removed the type packages, except `@types/wordpress__compose` which is not natively typed yet

Turns out most `@wordpress/*` are not really typed yet. For example, `@wordpress/data` has a `build-types` subfolder, but not a `types` field in `package.json` that would point to it. In the end, I was able to remove `@types/*` types only for `element`, `hooks` and `keycodes`.

noahtallen commented:
> @noahtallen @vindl What is a good way to test the upgrade here and can you possibly help? I don't really know what I'm doing when touching these parts of the codebase.

To get these changes on your wpcom sandbox, you have two options:
1. run `yarn dev --sync` from the `apps/editing-toolkit` directory. If you have `wpcom-sandbox` defined in your SSH config, it will compile and rsync the code to your sandbox whenever there are changes. Easiest for development, IMO.
2. Patch this diff to your sandbox: D55058-code (autogenerated above). This just takes a long time (several minutes) because of how big the diff is.

Then just sandbox a test site and the public API, and all the changes will appear when you visit the editor of that sandboxed site :)

I'm also happy to test these out!

jsnajdr committed:
Fix incorrect usage of i18n context in Gutenboarding

jsnajdr committed:
Use original colors where the color modifications were too disruptive

jsnajdr commented:
While testing the Editing Toolkit on my sandbox, I discovered the color changes (forced by `@wordpress/base-styles` deprecations) in the navigation sidebar are not very aesthetically pleasing:

Original:
<img width="274" alt="Screenshot 2021-01-12 at 9 06 58" src="https://user-images.githubusercontent.com/664258/104309095-e44bc100-54d1-11eb-9f5b-a1307fc91740.png">

Updated colors:
<img width="276" alt="Screenshot 2021-01-12 at 9 11 26" src="https://user-images.githubusercontent.com/664258/104309109-eca3fc00-54d1-11eb-9b13-613865960f7b.png">

The selected background color is too light. So I decided to use the original colors everywhere, and declare local variables with the values that are no longer in `@wordpress/base-styles`. Someone should update these colors to fit into some standardized palette. I don't feel qualified for that and it's off-topic for this PR, which aspires to be an upgrade that causes no visible changes.

jsnajdr commented:
@noahtallen In the end, the workflow that turned out to be the best fit for my need was the following. I needed to sync both `editing-toolkit` and `wpcom-block-editor` to my sandbox. I built both projects in production mode, and then used `yarn wpcom-sync` to sync them once with `rsync` to the sandbox host.

While doing that, I encountered some quirks that might be interesting to fix during your upcoming devex work 🙂 

The `yarn wpcom-sync` command in `wpcom-block-editor` doesn't do an initial sync. It only starts the `chokidar` watcher and syncs only after change. The same command in `editing-toolkit` behaves better: does an initial sync and only then starts watching.

I was interested only in syncing once, and was not interested in the watcher. There's no NPM command for that -- I had to modify the existing scripts to achieve that.

In `wpcom-block-editor`, running `yarn build:prod` builds the `dist` directory which contains also `*.asset.php` files. But these are not present in the wpcom's repo plugin folder. If these files are not needed, they shouldn't be built.

The `editing-toolkit` build could be much nicer, too. Right now it's many `dist` folders spread across the source tree, and we sync the entire folder, including sources, to `wpcom`. But we need only the `*.php` files and the `dist` folders with compiled JS and CSS. Building into a single `dist` folder and syncing only that would make for smaller and less bloated deploys.

Each script in the `editing-toolkit` is built separately, with its own webpack process and its own set of NPM commands. It would be nice to merge that into one webpack config with many entrypoints. Gutenberg build works exactly that way: all production scripts are part of one webpack compilation and are built by one command.

noahtallen commented:
Totally agree, @jsnajdr, thanks for outlining those. I'll add them to my list.

noahtallen commented:
> Each script in the editing-toolkit is built separately, with its own webpack process and its own set of NPM commands. It would be nice to merge that into one webpack config with many entrypoints. Gutenberg build works exactly that way: all production scripts are part of one webpack compilation and are built by one command.

The only drawback of this is that we do not always enqueue all of the script all of the time. For example, "dotcom-fse" (the legacy FSE experiment) assets do not need to be loaded most of the time, and currently they are not. That said, we could obviously do much better, especially for the things which are enqueued all the time.

jsnajdr commented:
> The only drawback of this is that we do not always enqueue all of the script all of the time.

That shouldn't be an obstacle at all. Gutenberg scripts are also not enqueued all the time. It's just about a streamlined one-step build process, the output artifacts should be exactly the same as before.

@noahtallen Could you please assist me with testing? I was able to verify that the patched plugins still load and work when synced to a wpcom sandbox, but I didn't test anything more sophisticated or subtle.

Also, does the D55058 synced diff look mergeable?

noahtallen reviewed (approved):
I spent a while smoke testing the various features, and nothing was broken for me. So this is good to merge 👍 

> Also, does the D55058 synced diff look mergeable?

We actually have a slightly different process -- you have to merge this PR first, and then create another PR to update the version. (Read more here: PCYsg-mMA-p2)

vindl commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nav-sidebar/src/wp-interface-types.d.ts:
```diff
@@ -1,3 +0,0 @@
-declare module '@wordpress/interface' {
```
Should we create a corresponding `wp-edit-post-types` file to carry this over for `@wordpress/edit-post`?

jsnajdr replied:
Oh yes, thanks for pointing this out. I originally removed this file because I assumed that all `@wordpress/*` packages now ship with types. But only some do. Now I addded a module definition for `@wordpress/edit-post` and it immediately fixed TypeScript errors on imports that were reported in the editor IDE.

vindl committed:
Restore __experimentalMainDashboardButton in expermental features e2e tests

vindl reviewed (commented):
Hey @jsnajdr, I pushed a commit to restore `__experimentalMainDashboardButton` in e2e tests (0621a12cffd7f7e8ba721e08d31d078556c45260), hope you don't mind.

I had a separate PR open to make the package changes https://github.com/Automattic/wp-calypso/pull/47702, but I'm going to close it in favor of this one, since it makes more sense to ship it simultaneously with monorepo update.

vindl commented:
I gave this a smoke test and it worked fine for me. 👍 There are some conflicts that need to be resolved before merging.

jsnajdr committed:
FSE: add type definitions for @wordpress/edit-post module

jsnajdr committed:
Final yarn.lock update

jsnajdr commented:
> I pushed a commit to restore `__experimentalMainDashboardButton` in e2e tests

Thanks @vindl ! Were the e2e tests failing because of the missing definitions?

vindl commented:
> Thanks @vindl ! Were the e2e tests failing because of the missing definitions?

They were supposed to by design, but I don't actually remember if they did before pushing that commit.

jsnajdr committed:
e2e: remove test for edit-site as we're testing on edit-post page

jsnajdr commented:
> I pushed a commit to restore `__experimentalMainDashboardButton` in e2e tests

@vindl I had to remove the `@wordpress/edit-site` check from the e2e test, because the test is running on a post editor page, in a `edit-post` context. The `edit-site` package is not loaded there and the check for it fails.

Is there such a thing as Full Site Editing embedded in Calypso Gutenframe? Doing the check on _that_ page would make it work.

jsnajdr committed:
Gutenboarding: use granular selectors from onboarding store

jsnajdr commented:
Now there is a Gutenboarding e2e test that fails reliably. Typing into the "acquire intent" box goes crazy:

<img width="978" alt="Screenshot 2021-01-20 at 14 02 45" src="https://user-images.githubusercontent.com/664258/105178499-623e4680-5b28-11eb-9bf6-664f435c39a2.png">

The input gets reset to an empty value several times while the webdriver is typing. It looks like some race condition when the e2e test does the typing and other actions very quickly.

vindl commented:
> Is there such a thing as Full Site Editing embedded in Calypso Gutenframe? Doing the check on that page would make it work.

@jsnajdr there is, but it's basically using the existing iframe post editor setup and just feeding it the different wp-admin url.

jsnajdr committed:
Disable the Gutenboarding e2e test

jsnajdr commented:
I disabled the Gutenboarding e2e test for now (cc @Automattic/luna). That made all checks green and I can finally merge this one (ufff).
